### PR TITLE
Histogram: Render heatmap-cells and heatmap-rows frames

### LINKED ISSUE #69790: Histogram: Should handle both classic and native histograms correctly for Prometheus datasource
### Discussed in https://github.com/grafana/grafana/discussions/69789

<div type='discussions-op-text'>

<sup>Originally posted by **krajorama** June  8, 2023</sup>
Histogram panel can display "classic" histograms from the Prometheus data source. The classic histograms were encoded as several time series representing cumulative buckets and a count and sum series.

Prometheus has introduced [native histograms](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms). This encodes sparse/exponential high resolution histograms into a single time series. This means the time series no longer contains floating point numbers, rather a complex data structure.

Result format on data source: https://prometheus.io/docs/prometheus/latest/querying/api/#native-histograms

For more information see presentations at PromCon for example: https://www.youtube.com/watch?v=RDdlDRkuVL8&t=2780s

image (6)

</div>

### PR DESCRIPTION
**Pull request by leeoniya**
Fixes https://github.com/grafana/grafana/issues/69790

[heatmaps-all.zip](https://github.com/grafana/grafana/files/13183731/heatmaps-all.zip)

before:
image

after:
image

yes, the legend is incorrect, but i'm gonna leave it for another PR, since it probably needs to touch a bunch of shared PlotLegend code.

image

with some drive-by fixes for `heatmap-cells` rendering in Heatmap.

1. we were 1-bucket off for axis ticks and y range was too small, so chopped off top bucket (it's hard to notice in the high resolution heatmap, but easy with the artificial one.
2. cell gap was also fixed to ensure centered alignment between tick boundaries.

before:
image

after:
image

### CODE REVIEW

File: public/app/plugins/panel/heatmap/utils.ts [hunk 1]
```diff
@@ -901,7 +907,7 @@ export function heatmapPathsSparse(opts: PathbuilderOpts) {
           ySize = Math.max(1, ySize - cellGap);
 
           let x = xMaxPx;
```
Comment (leeoniya) [inline public/app/plugins/panel/heatmap/utils.ts:909]: hmm, this probably needs to be xMin as well

**Review by ryantxu — APPROVED** [review]
LGTM  🌮 
image

File: public/app/plugins/panel/histogram/Histogram.tsx [hunk 2]
```diff
@@ -132,22 +154,24 @@ const prepConfig = (frame: DataFrame, theme: GrafanaTheme2) => {
...
-      const maxCount = u.bbox.width / ((maxWidth + labelSpacing) * devicePixelRatio);
-      const keepMod = Math.ceil(tickLabels.length / maxCount);
-
-      return tickLabels.map((label, i) => (i % keepMod === 0 ? label : null));
-    },
+    incrs: isOrdinalX ? [1] : useLogScale ? undefined : histogramBucketSizes,
+    splits: useLogScale || isOrdinalX ? undefined : xSplits,
+    values: isOrdinalX
+      ? (u: uPlot, splits: any[]) => splits
+      : (u: uPlot, splits: any[]) => {
```
Comment (ashharrison90) [inline public/app/plugins/panel/histogram/Histogram.tsx:161]: @leeoniya can we remove the explicit typing of `u` and `splits` here? they get inferred correctly as `uPlot` and `number[]` respectively 🥳

Comment (leeoniya) [reply public/app/plugins/panel/histogram/Histogram.tsx:161]: if TS is happy, then :+1:

**Pull request merged**